robot/impl/robot_reconfigure_weak_dependencies_test.go

```diff
@@ -285,4 +291,363 @@ func TestUpdateWeakDependents(t *testing.T) {
+	// (that also has an explicit dependency on one of the components).
+	// The following scenario is expected:
+	// 1) The two bases will configure first in parallel (which will each bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
+	//    checking the logical clock on also whether there are dependencies on weak dependents.
+	//    The check will fail.
+	// 3) weak1 will configure.
+	// 4) At the end of reconfiguration, weak1 will reconfigure once.
+	t.Log("Robot startup")
+	base1Name := base.Named("base1")
+	base2Name := base.Named("base2")
+	weakCfg1 := config.Config{
+		Components: []resource.Config{
+			{
+				Name:                weak1Name.Name,
+				API:                 weakAPI,
+				Model:               weakModel,
+				DependsOn:           []string{base1Name.String()},
+				ConvertedAttributes: &someTypeWithWeakAndStrongDepsConfig{},
+			},
+			{
+				Name:  base1Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+			{
+				Name:  base2Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+		},
+	}
+	test.That(t, weakCfg1.Ensure(false, logger), test.ShouldBeNil)
+	robot := setupLocalRobot(t, context.Background(), &weakCfg1, logger)
+
+	res, err := robot.ResourceByName(weak1Name)
+	// The resource was found and all dependencies were properly resolved.
+	test.That(t, err, test.ShouldBeNil)
+	weak1, err := resource.AsType[*someTypeWithWeakAndStrongDeps](res)
+	test.That(t, err, test.ShouldBeNil)
+	// Assert that the weak dependency was tracked.
+	test.That(t, weak1.resources, test.ShouldHaveLength, 2)
+	test.That(t, weak1.resources, test.ShouldContainKey, base1Name)
+	test.That(t, weak1.reconfigCount, test.ShouldEqual, 1)
+
+	// Reconfigure base1
+	// The following scenario is expected:
+	// 1) The base1 will reconfigure (which will bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
+	//    checking the logical clock on also whether there are dependencies on weak dependents.
+	//    The check will fail.
+	// 3) weak1 will not reconfigure, as base1 was not newly built or error during reconfiguration.
+	// 4) At the end of reconfiguration, weak1 will reconfigure once.
+	t.Log("Reconfigure base1")
+	weakCfg2 := config.Config{
+		Components: []resource.Config{
+			{
+				Name:                weak1Name.Name,
+				API:                 weakAPI,
+				Model:               weakModel,
+				DependsOn:           []string{base1Name.String()},
+				ConvertedAttributes: &someTypeWithWeakAndStrongDepsConfig{},
+			},
+			{
+				Name:  base1Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+				// We need the following `robot.Reconfigure` to call `Reconfigure` on this `base1`
+				// component. We change the `Attributes` field from the previous (nil) value to
+				// accomplish that.
+				Attributes: rutils.AttributeMap{"version": 1},
```

Comment:
I think ideally we'd express this small diff with a simpler:
```
weakCfg1.Components[1].Attributes = rutils...
```

Such that changing the `weakCfg1`s base2 or weak1 configs would not cause this reconfigure to do more than what's intended.

But I also know enough about configs in testing (looking at you ConvertedAttributes) to not assume it's always as simple as that. Is there a gotcha that my above suggestion falls into?

Reply:
I think this would work, too, but I have no strong opinion on what you currently have vs. @dgottlieb 's suggestion.

Reply:
updated